### PR TITLE
merge release/4.0.0 to dev

### PR DESCRIPTION
**Pull request by iamgusain**
Merging changes from release/4.0.0 back to dev

### CODE REVIEW

**Review by shahzaibj — APPROVED** [review]

**Review by AdamBJohnsonx — APPROVED** [review]
Need to cherry-pick more things back from master... but good.

File: common4j/src/test/com/microsoft/identity/common/java/platform/DeviceTest.java [hunk 1]
```diff
@@ -93,7 +93,7 @@ public void testGetOs(){
 
     @Test
     public void testGetProductVersion_DiagContextNotSet(){
-        Assert.assertEquals(NOT_SET, Device.getProductVersion());
+        Assert.assertEquals("1.5.9-default", Device.getProductVersion());
```
Comment (rpdome) [inline common4j/src/test/com/microsoft/identity/common/java/platform/DeviceTest.java:96]: This is probably an actual regression.

File: common4j/src/main/com/microsoft/identity/common/java/platform/Device.java [hunk 2]
```diff
@@ -101,8 +102,8 @@ public static String getProductVersion() {
 
         final String version = DiagnosticContext.INSTANCE.getRequestContext().get(AuthenticationConstants.SdkPlatformFields.VERSION);
         if (StringUtil.isNullOrEmpty(version)) {
-            Logger.warn(TAG + methodName, "Product version is not set.");
-            return NOT_SET;
+            Logger.warn(TAG + methodName, "Product version is not set.", null);
+            return StringUtil.isNullOrEmpty(BuildConfig.VERSION_NAME) ? "1.5.9-default" : BuildConfig.VERSION_NAME + "-default";
```
Comment (rpdome) [inline common4j/src/main/com/microsoft/identity/common/java/platform/Device.java:106]: Why are we hardcoding 1.5.9-default here?

NOT_SET means we're not initializing this properly. This version value is also platform-dependent. Not sure where 1.5.9 comes from.

Comment (iamgusain) [reply common4j/src/main/com/microsoft/identity/common/java/platform/Device.java:106]: This was changed in #1607

File: common4j/src/test/com/microsoft/identity/common/java/platform/DeviceTest.java (hunk 1 above)
Comment (iamgusain) [reply common4j/src/test/com/microsoft/identity/common/java/platform/DeviceTest.java:96]: This is due to the change in #1607

File: common4j/src/main/com/microsoft/identity/common/java/platform/Device.java (hunk 2 above)
Comment (rpdome) [reply common4j/src/main/com/microsoft/identity/common/java/platform/Device.java:106]: Let's add a comment + create an PBI to fix this properly.

**Review by rpdome — APPROVED** [review]

File: common4j/src/main/com/microsoft/identity/common/java/platform/Device.java (hunk 2 above)
Comment (rpdome) [reply common4j/src/main/com/microsoft/identity/common/java/platform/Device.java:106]: to 
1. undo the default value to NOT_SET (if we're getting this value, it means the upper layer didn't initialize the value properly).
2. Fix it in the upper layer(s).

**Pull request merged**